In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('replace-result.csv')
df

Unnamed: 0 Repository         File Name                 Class  Start Line  \
0         2931      tests  test_requests.py  Generator Expression        2184   
1         2932      tests  test_requests.py  Generator Expression        1016   
2         2933      tests  test_requests.py  Generator Expression        1412   
3         2934      tests  test_requests.py  Generator Expression        1419   
4         2935      tests  test_requests.py  Generator Expression        1449   
..         ...        ...               ...                   ...         ...   
56        5125   requests        cookies.py        Super Function         364   
57        7960   requests     exceptions.py        Super Function          24   
58        8015   requests     structures.py        Super Function          88   
59        8198   requests       adapters.py        Super Function         141   
60        8199   requests       adapters.py        Super Function         216   

    End Line  Displacement Level  \
0       2184            15    C1   
1       1016            33    C1   
2       1412            18    C1   
3       1419            18    C1   
4       1449            18    C1   
..       ...           ...   ...   
56       364            12    C2   
57        24             8    C2   
58        88             8    C2   
59       141             8    C2   
60       216             8    C2   

                                           Unnamed: 8  ChatGPT identify  \
0              data = (i for i in [b"a", b"b", b"c"])              True   
1           warnings_category = tuple(item.categor...              True   
2           assert all(isinstance(chunk, str) for ...              True   
3           assert all(isinstance(chunk, str) for ...              True   
4           assert all(len(chunk) == 1 for chunk i...              True   
..                                                ...               ...   
56                              super().update(other)              True   
57                  super().__init__(*args, **kwargs)              True   
58                                 super().__init__()              True   
59                                 super().__init__()              True   
60                                 super().__init__()              True   

    Code Change  Unnamed: 11  
0          True            0  
1          True            1  
2          True            0  
3          True            0  
4          True            0  
..          ...          ...  
56         True            1  
57         True            1  
58         True            1  
59         True            1  
60         True            1  

[61 rows x 12 columns]

## Generator Expression

In [4]:
df_genex = df[(df['Class']=="Generator Expression") & (df['Unnamed: 11'] == 0)]
df_genex

Unnamed: 0 Repository         File Name                 Class  Start Line  \
0         2931      tests  test_requests.py  Generator Expression        2184   
2         2933      tests  test_requests.py  Generator Expression        1412   
3         2934      tests  test_requests.py  Generator Expression        1419   
4         2935      tests  test_requests.py  Generator Expression        1449   
12        7513   requests          utils.py  Generator Expression         211   
13        7514   requests          utils.py  Generator Expression         791   

    End Line  Displacement Level  \
0       2184            15    C1   
2       1412            18    C1   
3       1419            18    C1   
4       1449            18    C1   
12       211            26    C1   
13       791            19    C1   

                                           Unnamed: 8  ChatGPT identify  \
0              data = (i for i in [b"a", b"b", b"c"])              True   
2           assert all(isinstance(chunk, str) for ...              True   
3           assert all(isinstance(chunk, str) for ...              True   
4           assert all(len(chunk) == 1 for chunk i...              True   
12          netrc_locations = (f"~/{f}" for f in N...              True   
13          no_proxy = (host for host in no_proxy....              True   

    Code Change  Unnamed: 11  
0          True            0  
2          True            0  
3          True            0  
4          True            0  
12         True            0  
13         True            0

2931: different types

2933: redo

2934: redo

2935: redo

7513: different types

7514: different types

## Generator Function (yield)

In [5]:
df_genfunc = df[(df['Class']=="Generator Function (yield)") & (df['Unnamed: 11'] == 0)]
df_genfunc

Unnamed: 0 Repository    File Name                       Class  \
16         724      tests  conftest.py  Generator Function (yield)   
17        4514      tests     utils.py  Generator Function (yield)   
18        5465   requests   cookies.py  Generator Function (yield)   
19        5466   requests   cookies.py  Generator Function (yield)   
20        5467   requests   cookies.py  Generator Function (yield)   
21        6250   requests  sessions.py  Generator Function (yield)   
22        6251   requests  sessions.py  Generator Function (yield)   
23        6996   requests    models.py  Generator Function (yield)   
24        6997   requests    models.py  Generator Function (yield)   
25        7955   requests     utils.py  Generator Function (yield)   
26        7956   requests     utils.py  Generator Function (yield)   
27        7957   requests     utils.py  Generator Function (yield)   
28        7958   requests     utils.py  Generator Function (yield)   
29        7959   requests     utils.py  Generator Function (yield)   

    Start Line  End Line  Displacement Level  \
16          55        55             4    C1   
17          14        14             8    C1   
18         232       232            12    C1   
19         249       249            12    C1   
20         266       266            12    C1   
21         263       263            16    C1   
22         280       280            16    C1   
23         888       888            12    C1   
24         835       835            20    C1   
25         578       578             8    C1   
26         587       587             8    C1   
27         756       756             8    C1   
28         311       311            12    C1   
29         575       575            12    C1   

                                           Unnamed: 8  ChatGPT identify  \
16      yield "localhost", server.server_address[1...              True   
17                                              yield              True   
18                                  yield cookie.name              True   
19                                 yield cookie.value              True   
20                    yield cookie.name, cookie.value              True   
21                                          yield req              True   
22                                         yield resp              True   
23                                        yield chunk              True   
24                                      yield pending              True   
25                                           yield rv              True   
26             yield string[pos : pos + slice_length]              True   
27                                              yield              True   
28                                           yield rv              True   
29                                  yield tmp_handler              True   

    Code Change  Unnamed: 11  
16         True            0  
17         True            0  
18         True            0  
19         True            0  
20         True            0  
21         True            0  
22         True            0  
23         True            0  
24         True            0  
25         True            0  
26         True            0  
27         True            0  
28         True            0  
29         True            0

In [8]:
file_path = 'result/ori-replaced-edited.csv'
data = pd.read_csv(file_path)
replacement_counts = data.groupby(['Class', 'Replaced Element']).size().reset_index(name='Count')

most_common_replacements = replacement_counts.sort_values(by='Count', ascending=False)
most_common_replacements

Class      Replaced Element  Count
3         Generator Expression     Simple Assignment     16
4         Generator Expression      Simple Atributte     13
7         Generator Expression           Simple List     13
5         Generator Expression       Simple For Loop     12
14  Generator Function (yield)                Return     10
..                         ...                   ...    ...
31           Import're' module  Generator Expression      1
32           Import're' module                Return      1
33           Import're' module     Simple Assignment      1
34           Import're' module  Simple If statements      1
60          map' call function           Simple List      1

[61 rows x 3 columns]

In [9]:
common_genfunc = most_common_replacements[most_common_replacements['Class']=="Generator Function (yield)"]
common_genfunc

Class                 Replaced Element  Count
14  Generator Function (yield)                           Return     10
17  Generator Function (yield)                  Simple For Loop      5
18  Generator Function (yield)                      Simple List      4
16  Generator Function (yield)                 Simple Atributte      4
15  Generator Function (yield)                Simple Assignment      3
13  Generator Function (yield)       Generator Function (yield)      3
19  Generator Function (yield)                     Simple Tuple      2
12  Generator Function (yield)  For Loop with 1 List to iterate      1
20  Generator Function (yield)                  pass' statement      1